### PR TITLE
BREAKING CHANGE: Implement conversation-based any2t pipeline

### PR DESCRIPTION
**Pull request by Lavezlyn**
- Support MMMU benchmark.
- **All t2t benchmarks work well after refactoring.**
- Refactor `InferenceInput`, replace _text_ field with _conversation_, remove redundant fields.
- Refactor `InferenceOutput`, remove redundant fields.
- Rewrite t2t inference logic in `vllm_lm.py`, use `apply_chat_template` method to simplify inference based on conversation.
- Support ti2t inference in `vllm_mm.py`.
- Add necessary multi-modal data processing helper functions in `utils/mm_data_manager`.
- Refine development doc.

TODO:
- [ ] Generalize inference pipeline in `vllm_mm.py`, currently works only with text-image input. 
- [ ] Add conversation field to `InferenceOutput` to better support multi-modal generation tasks.

### CODE REVIEW

File: eval_anything/dataloader/t2t_dataloader.py [hunk 1]
```diff
@@ -27,10 +27,26 @@
...
 class T2TDataLoader(BaseDataLoader):
+
+    def build_conversation_from_prompt(self, prompt: str):
+        """
+        Build a default conversation from a text prompt.
+        """
+        conversation = [
+            {
+                "role": "system",
+                "content": "You are a helpful assistant."
```
Comment (XuyaoWang) [inline eval_anything/dataloader/t2t_dataloader.py:38]: You cannot explicitly add the system prompt, as different models have different system prompts. The logic for system prompts will be handled by the `apply_chat_template` function.

File: eval_anything/benchmarks/text_image_to_text/mmmu/configs.yaml [hunk 2]
```diff
@@ -22,34 +22,118 @@ dataset:
...
   cot_fewshot_data_split: null
   max_shot: 5
-  default_task_list: ["Accounting"]
+  default_task_list: ["Accounting", "Agriculture", "Architecture_and_Engineering", "Art", "Art_Theory", "Basic_Medical_Science", "Biology", "Chemistry", "Clinical_Medicine", "Computer_Science", "Design", "Diagnostics_and_Laboratory_Medicine", "Economics", "Electronics", "Energy_and_Power", "Finance", "Geography", "History", "Literature", "Manage", "Marketing", "Materials", "Math", "Mechanical_Engineering", "Music", "Pharmacy", "Physics", "Psychology", "Public_Health", "Sociology"]
+task_defaults: &task_defaults
+  type: MultiChoice
+  question_key: question
+  answer_key: options
+  ground_truth_key: answer
+  candidate_labels: ["A", "B", "C", "D"]
```
Comment (XuyaoWang) [inline eval_anything/benchmarks/text_image_to_text/mmmu/configs.yaml:31]: MMMU's options are variable; they could be one of [A,B], [A,B,C], [A,B,C,D], or [A,B,C,D,E].

File: eval_anything/models/vllm_lm.py [hunk 3]
```diff
@@ -64,35 +63,39 @@ def init_model(self) -> None:
...
-                return_tensors="pt"
-            )
-            prompts = [
-                self.tokenizer.decode(input_id, skip_special_tokens=True)
-                for input_id in input_ids
-            ]
+        input_list = self._build_conversation_from_template(input_list)
+        
+        prompts_token_ids = self.tokenizer.apply_chat_template(
+            [input.conversation for input in input_list],
```
Comment (XuyaoWang) [inline eval_anything/models/vllm_lm.py:87]: In the `apply_chat_template` function, add `tokenize=False` to get the input string with the template added as a string type, without needing to use the decode function.

File: eval_anything/benchmarks/text_image_to_text/mmmu/configs.yaml (hunk 2 above)
Comment (Lavezlyn) [reply eval_anything/benchmarks/text_image_to_text/mmmu/configs.yaml:31]: This can be solved by setting candidate_labels to ["A","B","C","D","E"] and the multi-choice prompt builder can handle the varying length automatically.

File: eval_anything/evaluate_tools/t2t_tools.py [hunk 4]
```diff
@@ -197,6 +247,154 @@ def match_text(text):
...
         matches = [match_text(item) for item in data]
         return matches
+
+class RegexMatchMultiOpen(RegexMatch):
+    """
+    Handle multi-choice and open-ended mixed tasks.
+    """
+    def __init__(self, additional_pattern: str = None, match_index: int = None):
+        super().__init__(additional_pattern, match_index)
+        self.letter_pattern = r"\b([A-D])\b|\((?P<letter>[A-D])\)"
```
Comment (Kass123777) [inline eval_anything/evaluate_tools/t2t_tools.py:257]: The `letter_pattern` is supposed to be more flexible, not just from 'A' to 'D'.

**Comment by Kass123777** [comment]
Have you tested all the tasks in the MMMU benchmark?

**Comment by Lavezlyn** [comment]
Have tested all tasks. Llava-1.5-7b-hf achieves 43.3% accuracy on average.

File: eval_anything/evaluate_tools/t2t_tools.py (hunk 4 above)
Comment (Lavezlyn) [reply eval_anything/evaluate_tools/t2t_tools.py]: fix this in the latest commit, add an optional argument so that user can set letter_pattern in config.yaml

**Review by Kass123777 — APPROVED** [review]
Great work!
